### ISSUE TITLE
Proposal: new release of DataFusion?

### ISSUE THREAD

**Issue by alamb** [apache/arrow-datafusion#1587]
I wonder if it is time to release a new version of datafusion to crates.io? 

It would be great to crowdsource:
1. Update readme / changelog
2. Update version
2. (maybe) a blog post?

I am happy to handle creating a release candidate / doing the official voting process.

**Comment by xudong963** [comment]
Yeah, I think we can wait for arrow2 related tickets merged into master? 
BTW, I can help write a blog!

**Comment by alamb** [comment]
arrow2 may be a good driver. 

I don't have a good sense of how many projects use datafusion from `crates.io` (aka what has been released) vs how many use it via a `github` sha . IOx (my project) [uses the sha](https://github.com/influxdata/influxdb_iox/blob/main/datafusion/Cargo.toml#L12) but I realized that maybe others are waiting for an actual release

**Comment by alamb** [comment]
There is all sorts of good stuff in DataFusion since we last made a release. Since arrow (C++) just released a 7.0.0 I was thinking to do the same with DataFusion (as @pauldix  says the success of a project is predicated on 1. Making sweet sweet software and 2. telling people about it). We have done 1, and now we need to do a bit more of 2. ✍️

**Comment by matthewmturner** [comment]
Perhaps we can start crowdsourcing points for a post / blog on a google doc.  I've made one here (sry havent had chance to add anything yet but trying to help as i can):

https://docs.google.com/document/d/17uB1GIN58xOehQP5XpJH8J3qA7KDS1WQJzyHzW7L1kU/edit?usp=sharing

**Comment by houqp** [comment]
I won't have the time to drive the release this time, but happy to help fixing any issue with the existing release automation and guide anyone through the process.

**Comment by alamb** [comment]
I can drive the release

**Comment by xudong963** [comment]
I will finish this issue https://github.com/apache/arrow-datafusion/issues/1400 before release to give our users a clearer readme

**Comment by alamb** [comment]
Thanks @matthewmturner  -- I'll try and give the doc another pass through later today

**Comment by matthewmturner** [comment]
I went through and made more updates and added some git stats.

@alamb let me know if anything in particular I can do to help alleviate the burden on your side.  I'm happy to provide more assistance on releases in general (this one and future).

**Comment by alamb** [comment]
Thanks you very much @matthewmturner 

Things that I think we need to do prior to release:
- [ ] Make a PR to update the version in all the crates to 7.0.0
- [ ] Create a PR to create a CHANGELOG (I had good luck with the github changelog generator that @jorgecarleitao  setup in arrow-rs once upon a time). 
- [ ] (nice to have): https://github.com/apache/arrow-datafusion/issues/1741 
- [ ] Take a pass through the docs site to refresh / improve it / make it "snazzy"


Making the PR to update the version would be sweet 👍 

I think it is probably best if I take an initial swag at running the changelog generator thing as it requires the ability to mess with tickets titles / tags.

**Comment by alamb** [comment]
Here is a proposed changelog. https://github.com/apache/arrow-datafusion/pull/1807#

I guess it is probably time to cut a release branch. Does anyone know if we are waiting for anything else to be merged? @xudong963  @matthewmturner @Dandandan @houqp @Dandandan @Jimexist @andygrove ?

**Comment by matthewmturner** [comment]
@HaoYang670 did you get a chance to work on #1741?  

@alamb its probably too late to have any impact for this release, but for my info, can we only do updates to the user guide / website (https://arrow.apache.org/datafusion/) when we have a release?  I figure any docs.rs update will have to be linked to a release.

Otherwise, okay for me.

**Comment by alamb** [comment]
>  can we only do updates to the user guide / website (https://arrow.apache.org/datafusion/) when we have a release?

No, we can update https://arrow.apache.org/datafusion/ (the hosted version of https://github.com/apache/arrow-datafusion/tree/master/docs) any time we want 👍

**Comment by xudong963** [comment]
> I guess it is probably time to cut a release branch. Does anyone know if we are waiting for anything else to be merged?

I take a quick look at our unmerged tickets, I think no. Thanks for your nice work! @alamb

**Comment by HaoYang670** [comment]
I am working on https://github.com/apache/arrow-datafusion/issues/1741
these days. And I will file a PR today.

**Comment by alamb** [comment]
I am going to do a dry run today of publishing datafusion to crates.io (will use an 0.1.xx version to test)

**Comment by alamb** [comment]
I tested publishing `7.0.0-alpha` to crates.io using https://github.com/alamb/arrow-datafusion/tree/alamb/test_publish

then going into each crate like this:

 `cd datafusion-common && cargo publish`

* https://crates.io/crates/datafusion-common/7.0.0-alpha
* https://crates.io/crates/datafusion-expr/7.0.0-alpha
* https://crates.io/crates/datafusion/7.0.0-alpha

One thing I noticed is that `datafusion-cli` depends on ballista, and so without a ballista release we can't also do a datafusion-cli release 🤔  But maybe that is ok and we can do a datafusion-cli release later (or maybe even move datafusion cli into the contrib repo, or perhaps make the ballista dependency optional)

**Comment by matthewmturner** [comment]
I had thought about moving `datafusion-cli` to datafusion-contrib as well.  This does seem generally aligned with moving things out of the `datafusion` repo that arent actually `datafusion` (i.e. `datafusion-python`).  It does also seem reasonable to make `ballista` optional.

**Comment by matthewmturner** [comment]
@Jimexist That's a really good point - I have found it helpful as well. But couldn't we do something like the below in 'cargo.toml' from the datafusion-cli repo to achieve a similar experience? Not quite as convenient but I think it's close. 

'''
[dependencies]
datafusion = { path = "../path/to/datafusion" }
'''

**Comment by alamb** [comment]
> Having a cli in repo helps a lot with debugging in my opinion

This is an excellent point @Jimexist  -- I use the cli extensively during debugging. 

> But couldn't we do something like the below in cargo.toml from the datafusion-cli repo to achieve a similar experience? Not quite as convenient but I think it's close.

I think we could @matthewmturner  but I think that will effectively require people to have checked out the datafusion-cli repo any time they want to build datafusion; So if it is required to build datafusion,  the benefits of putting it in a separate repo seem pretty small 🤔

**Comment by alamb** [comment]
Filed https://github.com/apache/arrow-datafusion/issues/1814 to see if we can solict some more help for the doc site

**Comment by alamb** [comment]
https://github.com/apache/arrow-datafusion/pull/1816 <-- for optional ballista feature in `datafusion-cli`

**Comment by alamb** [comment]
🤔  I don't think  https://github.com/apache/arrow-datafusion/pull/1816 is sufficient to publish `datafusion-cli` to crates.io -- cargo still tries to resolve the dependencies of ballista (even though it is an optional dependency)

The only way I could get it to publish was to comment out the ballista dependency all together 8750db3bea 🤔

**Comment by alamb** [comment]
Update here: I would like to wait for the arrow 9.0.0 release to be published (later today or tomorrow) and then update datafusion to use it: https://github.com/apache/arrow-datafusion/pull/1775

Then I'll try and make a release candidate tomorrow or Monday

Its going to be great!

**Comment by alamb** [comment]
I am actively working to create a release candidate for datafusion 7.0.0

**Comment by alamb** [comment]
@matthewmturner would you be willing to start a PR for a blog post:   https://docs.google.com/document/d/17uB1GIN58xOehQP5XpJH8J3qA7KDS1WQJzyHzW7L1kU/edit?usp=sharing ?

The DataFusion 6.0.0 announcement is here for reference:  https://github.com/apache/arrow-site/pull/160

**Comment by matthewmturner** [comment]
> @matthewmturner would you be willing to start a PR for a blog post: https://docs.google.com/document/d/17uB1GIN58xOehQP5XpJH8J3qA7KDS1WQJzyHzW7L1kU/edit?usp=sharing ?
> 
> The DataFusion 6.0.0 announcement is here for reference: [apache/arrow-site#160](https://github.com/apache/arrow-site/pull/160)

Sure - will do now!

**Comment by alamb** [comment]
Official mailing list post with all the details is here: https://lists.apache.org/thread/t8381y8x1t452dvqr3y7h85q4dncvwrx

**Comment by xudong963** [comment]
Thanks, @alamb @matthewmturner <https://github.com/matthewmturner>!


Matthew Turner ***@***.***> 于2022年2月14日周一 22:58写道：

**Comment by alamb** [comment]
The release was approved and published 🎉 

Mailing list thread is here: https://lists.apache.org/thread/hcpcf3shlt0l3fm2k313tq1tvrczlowf

The release is available here:
  https://dist.apache.org/repos/dist/release/arrow/arrow-datafusion-7.0.0

I have also published it to crates.io here:
https://crates.io/crates/datafusion/7.0.0
